### PR TITLE
Add mode="eager-numerics" and accuracy_high sugar to torch.compile

### LINKED ISSUE #189334: [Feature Request] Add accuracy_high=True or mode="max-precision" to torch.compile
### New Feature for Release

Aggressive graph fusions and backend math optimizations in torch.compile can introduce minor numerical drift compared to eager mode. Currently, forcing maximum precision requires deep knowledge of internal compiler flags via the options dictionary. There is no simple, high-level flag for users who need strict numerical fidelity over maximum speed.

#Proposed API

opt_model = torch.compile(model, accuracy_high=True)

# OR
opt_model = torch.compile(model, mode="max-precision")
Proposed Behavior
Activating this would automatically bundle safe configurations under the hood:

Enforcing strict IEEE-754 math compliance in backend code generation (Triton/Inductor).

Disabling or bounding lossy floating-point reordering during optimization fusions.

### Point(s) of contact

_No response_

### Release Mode (pytorch/pytorch features only)

In-tree

### Out-Of-Tree Repo

_No response_

### Description and value to the user

_No response_

### Link to design doc, GitHub issues, past submissions, etc

_No response_

### What feedback adopters have provided

_No response_

### Plan for documentations / tutorials

Tutorial exists

### Additional context for tutorials

_No response_

### Marketing/Blog Coverage

Yes

### Are you requesting other marketing assistance with this feature?

_No response_

### Release Version

_No response_

### OS / Platform / Compute Coverage

_No response_

### Testing Support (CI, test cases, etc..)

_No response_

cc @chauhang @penguinwu @malfet

### PR DESCRIPTION
**Pull request by heytanix**
Fixes #189334.

## Summary

Achieving eager-like numerics in torch.compile today requires knowing about scattered Inductor config knobs (emulate_precision_casts, eager_numerics.use_pytorch_libdevice, use_fast_math, per backend fast-math C++ flags, etc). There was no single high-level switch, so users who need strict numerical fidelity over speed had to discover and combine these internal flags manually.

This adds "eager-numerics" as a new entry in torch._inductor.list_mode_options(), the existing extensibility point used by "reduce-overhead" and "max-autotune". It bundles the knobs that keep Inductor's codegen closer to eager: preserving fp16/bf16 downcast-upcast pairs across fused ops instead of eliding them, using the CUDA toolkit's libdevice for transcendental functions instead of Triton's bundled version, and disabling FTZ and fast-math codegen paths (CPU, ROCm, CUTLASS). accuracy_high=True is added to torch.compile() as sugar that maps to mode="eager-numerics", matching the two API shapes proposed in the issue; it raises RuntimeError if combined with mode or options, mirroring the existing mode/options mutual exclusion check.

Note this is deliberately named "eager-numerics" rather than "max-precision" (see PR discussion below): some of the bundled flags, like emulate_precision_casts, reduce precision relative to what Inductor would otherwise compute, they just make the result closer to eager's rounding behavior. "Precision" and "matches eager" are not the same axis.

TF32 and matmul precision (torch.backends.cuda.matmul.fp32_precision, cudnn, mkldnn) are global process-wide flags, not torch._inductor.config keys, so they cannot be set through the mode's config dict the way the other knobs are. Rather than mutating global backend state as a side effect of compiling one function, the new mode leaves them alone and the docstring points users at torch.set_float32_matmul_precision("highest") and torch.backends.cudnn.allow_tf32 = False instead.

LLM used: Claude Code

## Test plan

New tests in test/inductor/test_config.py:
```
python test/inductor/test_config.py -k test_api_options
python test/inductor/test_config.py -k test_accuracy_high
```

Full existing suite to check for regressions:
```
python test/inductor/test_config.py
```

Manual sanity check:
```python
import torch
print(torch._inductor.list_mode_options('eager-numerics'))

@torch.compile(accuracy_high=True)
def f(x):
    return torch.sin(x) + torch.cos(x)
print(f(torch.randn(8)))
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by heytanix** [comment]
The `distributed` test shard 2/8 failure looks like a runner infra issue (pod stuck in `Pending` - `CreateContainerError` - backoff timeout after 2h, no test code executed), unrelated to this PR's changes. Could someone with write access re-run that failed job when you get a chance? Thanks!

**Comment by jbschlosser** [comment]
Hi @heytanix, thanks for the contribution! Note that we require consensus on the associated issue before accepting a PR like this. Just a heads-up that compile folks will need to weigh in before this is accepted / reviewed :)

**Review by eellison — COMMENTED** [review]
cc @karthickai . i think we probably want to deliberate on the api more before accepting this.

**Comment by karthickai** [comment]
> cc @karthickai . i think we probably want to deliberate on the api more before accepting this.

@heytanix there's already active work on eager and torch.compile bitwise equivalence under a dedicated strict-numerics flag, `torch._inductor.config.numerics = "strict"` (proposed in #187971) check my [comment here](https://github.com/pytorch/pytorch/issues/189334#issuecomment-4975046833). That effort is currently focused on reduction ordering (the main drift source). What I can do  put up a separate PR for the strict-numerics API on its own, and you can use that group it together
cc @eellison @jansel

**Comment by heytanix** [comment]
Thanks for the context @karthickai, and for flagging this @eellison.

Good to know `strict_numerics` is in flight. Looking at the diff in #187971, it looks like our two pieces are actually complementary rather than overlapping: `numerics="strict"` targets reduction accumulation order (sum, INNER_TREE), while this PR's `max-precision` mode targets a different set of drift sources - fp16/bf16 downcast-upcast elision across fusions, transcendental function precision (libdevice vs Triton's bundled version), FTZ, division rounding, and fast-math compiler flags (C++/ROCm/CUTLASS). There's no literal code overlap today.

That said, I agree it makes sense to consolidate rather than end up with two independently-named "make numerics stricter" knobs. Happy to hold this PR until `strict_numerics` lands, then update `mode="max-precision"` / `accuracy_high` to also set `config.numerics="strict"` alongside the existing bundle - so the mode becomes the single high-level entry point, and `config.numerics` stays the lower-level building block you're expanding over time. I'll note the caveat that `numerics="strict"` currently no-ops silently on Triton builds without `tl.ReductionOrdering`, so that should probably be surfaced somewhere in the mode's docs too.

Let me know if that grouping makes sense, or if you'd rather keep them fully separate (e.g. mode stays scoped to what it covers today, and users opt into `numerics="strict"` independently). Happy to adjust either way once #187971 merges.

**Comment by powerofaisinstudy-debug** [comment]
Thank you so much for the quick updates, @karthickai! It’s awesome to see the backend work on the strict flag in #187971.

I also saw that the Pull Request #189422 was opened to add the mode="max-precision" and accuracy_high syntax sugar directly inspired by this issue. Thank you to the author for putting that together so fast!

I’d love to help collaborate and get this merged. I am happy to help test the new max-precision flag on my own deep learning models to verify that the numerical drift is completely resolved. Let me know if there are specific edge cases you want me to run, or if I can help look into any of the failing CI checks!

**Review by jansel — CHANGES_REQUESTED** [review]
"max-precision" is the wrong name for some of these flags.  For example `"emulate_precision_casts": True` *reduces* precision (but is closer to what eager does).

As others have said, there is work in progress on a stricter eager-matching numerics mode.

**Comment by powerofaisinstudy-debug** [comment]
Thanks for the feedback That makes total sense regarding emulate_precision_casts. Since the ultimate goal of these combined configurations is achieving strict numerical fidelity to match eager mode exactly, what do you think about renaming it to mode="eager-numerics" instead?

cc @eellison @karthickai @jansel

**Comment by powerofaisinstudy-debug** [comment]
Hi @heytanix, I’ve been tracking the PR status and noticed the linux-jammy-py3.10-gcc11 check is failing consistently on container initialization. It looks like an infrastructure/runner issue rather than a code problem. Would you like me to ping a maintainer about the infra failure, or would you prefer to handle the PR updates from your side?

**Comment by heytanix** [comment]
Renamed to `mode="eager-numerics"` in the latest commit - agreed that "max-precision" mischaracterized flags like `emulate_precision_casts` (which reduce precision to match eager's rounding, not increase it). Also expanded the docstring to call out that distinction explicitly.

`accuracy_high=True` now maps to `mode="eager-numerics"` accordingly. Let me know if this naming works, or if there's a preference to reconcile it further with `config.numerics` once #187971 lands.

**Comment by heytanix** [comment]
Thanks for keeping an eye on it, @powerofaisinstudy-debug! Already flagged that one in an earlier comment above - it's a runner pod `CreateContainerError` that times out, not a real test failure, so no action needed on the code side. No need to ping separately, but appreciate you double-checking.

And thanks for offering to help test - once the naming settles (see the `eager-numerics` rename above) I'll ping you here if there are good edge cases to try on real models.

**Comment by karthickai** [comment]
Hi @heytanix, as I mentioned before, since this is still experimental, I think it's safer to expose it behind explicit config flags first. Could you hold until the config.numerics PR lands?

And before we group anything into a mode, we should first pin down where the numeric differences actually come from (we're actively fixing reductions) and how the grouping affects Inductor perf i.e. more tests + ci perf benches so we understand the impact before committing to a mode.

**Comment by heytanix** [comment]
Yes @karthickai,
I have just made the changes that @jansel had requested,
As said earlier, I'll keep this on hold.

**Comment by powerofaisinstudy-debug** [comment]
Agreed. Moving this behind config flags is the right approach to ensure stability while we investigate the numerical differences. I'm happy to help verify performance benchmarks or test edge cases once the config infrastructure is ready, just let me know if you need an extra pair of eyes.

**Comment by powerofaisinstudy-debug** [comment]
Hey  @karthickai, @heytanix,  @jansel, @karthickai 

Following up on the reduction discussions—here is a quick, self-contained MRE isolating numerical drift during high-dimensional reductions between Eager mode and Inductor:

```python
import torch

device = "cuda" if torch.cuda.is_available() else "cpu"
x = torch.randn(2048, 2048, dtype=torch.float32, device=device)

eager_out = torch.sum(x, dim=-1)
compiled_fn = torch.compile(lambda t: torch.sum(t, dim=-1))
compiled_out = compiled_fn(x)

max_diff = torch.max(torch.abs(eager_out - compiled_out)).item()
print(f"Max absolute divergence: {max_diff:.6e}")
